# CASE STUDY: CUSTOMER ANALYSIS FOR RETAIL

## ANALYTICS IN RETAIL:

With the retail market getting more and more competitive by the day, there has never been anything more important than the ability for optimizing service business processes when trying to satisfy the expectations of customers. Channelizing and managing data with the aim of working in favor of the customer as well as generating profits is very significant for
survival.
    Ideally, a retailer’s customer data reflects the company’s success in reaching and nurturing its customers. Retailers built reports summarizing customer behavior using metrics such as conversion rate, average order value, recency of purchase and total amount spent in recent transactions. These measurements provided general insight into the behavioral tendencies
of customers.
    Customer intelligence is the practice of determining and delivering data-driven insights into past and predicted future customer behavior.To be effective, customer intelligence must combine raw transactional and behavioral data to generate derived measures.
    In a nutshell, for big retail players all over the world, data analytics is applied more these days at all stages of the retail process – taking track of popular products that are emerging, doing forecasts of sales and future demand via predictive simulation, optimizing placements of products and offers through heat-mapping of customers and many others.
    
## DATA AVAILABILITY:

### Retail Data.xlsx

    o This book has three sheets (Customer, Transaction, Product Heirarchy)
    o Customer: Customers information including demographics
    o Transaction: Transactions of customers
    o Product Heirarchy: Product information (cateogry, sub category etc...)
    
## BUSINESS PROBLEM:

A Retail store is required to analyze the day-to-day transactions and keep a track of its customers spread across various locations along with their purchases/returns across various categories. 

Create a report and display the below calculated metrics, reports and inferences.

    1. Merge the datasets Customers, Product Hierarchy and Transactions as Customer_Final. Ensure to keep all customers who have done transactions with us and select the join type accordingly.
    2. Prepare a summary report for the merged data set.
    
        a. Get the column names and their corresponding data types
        b. Top/Bottom 10 observations
        c. “Five-number summary” for continuous variables (min, Q1, median, Q3 and max)
        d. Frequency tables for all the categorical variables
    3. Generate histograms for all continuous variables and frequency bars for categorical variables.
    4. Calculate the following information using the merged dataset :
    
        a. Time period of the available transaction data
        b. Count of transactions where the total amount of transaction was negative
    5. Analyze which product categories are more popular among females vs male customers.
    6. Which City code has the maximum customers and what was the percentage of customers from
    that city?
    7. Which store type sells the maximum products by value and by quantity?
    8. What was the total amount earned from the "Electronics" and "Clothing" categories from
    Flagship Stores?
    9. What was the total amount earned from "Male" customers under the "Electronics" category?
    10. How many customers have more than 10 unique transactions, after removing all transactions
    which have any negative amounts?
    11. For all customers aged between 25 - 35, find out:
    
        a. What was the total amount spent for “Electronics” and “Books” product categories?
        b. What was the total amount spent by these customers between 1st Jan, 2014 to 1st Mar, 2014?
        
---
---

# Import Libraries

In [192]:
# Handle table-like data and matrices
import pandas as pd
import numpy as np
from numpy import percentile
from datetime import date

# Modelling Helpers
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Visualisation
import matplotlib.pyplot as plt
import seaborn as sns
import cufflinks as cf
import plotly
import plotly.express as px
from plotly import __version__
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
init_notebook_mode(connected = True)
cf.go_offline()

# Configure visualisations
%matplotlib inline

# Load Data

In [3]:
customer_data= pd.read_csv('Customer.csv')
customer_data

customer_Id         DOB Gender  city_code
0          268408  02-01-1970      M        4.0
1          269696  07-01-1970      F        8.0
2          268159  08-01-1970      F        8.0
3          270181  10-01-1970      F        2.0
4          268073  11-01-1970      M        1.0
...           ...         ...    ...        ...
5642       274474  19-12-1992      M        2.0
5643       267666  24-12-1992      M        6.0
5644       270476  25-12-1992      F        3.0
5645       269626  27-12-1992      F        5.0
5646       274308  29-12-1992      F        5.0

[5647 rows x 4 columns]

In [4]:
prod_cat_data= pd.read_csv('prod_cat_info.csv')
prod_cat_data

prod_cat_code          prod_cat  prod_sub_cat_code          prod_subcat
0               1          Clothing                  4                 Mens
1               1          Clothing                  1                Women
2               1          Clothing                  3                 Kids
3               2          Footwear                  1                 Mens
4               2          Footwear                  3                Women
5               2          Footwear                  4                 Kids
6               3       Electronics                  4              Mobiles
7               3       Electronics                  5            Computers
8               3       Electronics                  8  Personal Appliances
9               3       Electronics                  9              Cameras
10              3       Electronics                 10      Audio and video
11              4              Bags                  1                 Mens
12              4              Bags                  4                Women
13              5             Books                  7              Fiction
14              5             Books                 12             Academic
15              5             Books                 10          Non-Fiction
16              5             Books                 11             Children
17              5             Books                  3               Comics
18              5             Books                  6                  DIY
19              6  Home and kitchen                  2           Furnishing
20              6  Home and kitchen                 10              Kitchen
21              6  Home and kitchen                 11                 Bath
22              6  Home and kitchen                 12                Tools

In [5]:
transaction_data= pd.read_csv('Transactions.csv')
transaction_data

transaction_id  cust_id   tran_date  prod_subcat_code  prod_cat_code  \
0         80712190438   270351  28-02-2014                 1              1   
1         29258453508   270384  27-02-2014                 5              3   
2         51750724947   273420  24-02-2014                 6              5   
3         93274880719   271509  24-02-2014                11              6   
4         51750724947   273420  23-02-2014                 6              5   
...               ...      ...         ...               ...            ...   
23048     94340757522   274550  25-01-2011                12              5   
23049     89780862956   270022  25-01-2011                 4              1   
23050     85115299378   271020  25-01-2011                 2              6   
23051     72870271171   270911  25-01-2011                11              5   
23052     77960931771   271961  25-01-2011                11              5   

       Qty  Rate      Tax  total_amt Store_type  
0       -5  -772  405.300  -4265.300     e-Shop  
1       -5 -1497  785.925  -8270.925     e-Shop  
2       -2  -791  166.110  -1748.110   TeleShop  
3       -3 -1363  429.345  -4518.345     e-Shop  
4       -2  -791  166.110  -1748.110   TeleShop  
...    ...   ...      ...        ...        ...  
23048    1  1264  132.720   1396.720     e-Shop  
23049    1   677   71.085    748.085     e-Shop  
23050    4  1052  441.840   4649.840        MBR  
23051    3  1142  359.730   3785.730   TeleShop  
23052    1   447   46.935    493.935   TeleShop  

[23053 rows x 10 columns]

In [41]:
prod_cat_data.rename(columns={"prod_sub_cat_code":"prod_subcat_code"},inplace=True)

# BUSINESS PROBLEM:

A Retail store is required to analyze the day-to-day transactions and keep a track of its customers spread across various locations along with their purchases/returns across various categories.

Create a report and display the below calculated metrics, reports and inferences.

**1. Merge the datasets Customers, Product Hierarchy and Transactions as Customer_Final. Ensure to keep all customers who have done transactions with us and select the join type accordingly.**

In [42]:
tra_prod_data = pd.merge(left=transaction_data, right=prod_cat_data,on=["prod_cat_code","prod_subcat_code"],how="left")
tra_prod_data

transaction_id  cust_id   tran_date  prod_subcat_code  prod_cat_code  \
0         80712190438   270351  28-02-2014                 1              1   
1         29258453508   270384  27-02-2014                 5              3   
2         51750724947   273420  24-02-2014                 6              5   
3         93274880719   271509  24-02-2014                11              6   
4         51750724947   273420  23-02-2014                 6              5   
...               ...      ...         ...               ...            ...   
23048     94340757522   274550  25-01-2011                12              5   
23049     89780862956   270022  25-01-2011                 4              1   
23050     85115299378   271020  25-01-2011                 2              6   
23051     72870271171   270911  25-01-2011                11              5   
23052     77960931771   271961  25-01-2011                11              5   

       Qty  Rate      Tax  total_amt Store_type          prod_cat prod_subcat  
0       -5  -772  405.300  -4265.300     e-Shop          Clothing       Women  
1       -5 -1497  785.925  -8270.925     e-Shop       Electronics   Computers  
2       -2  -791  166.110  -1748.110   TeleShop             Books         DIY  
3       -3 -1363  429.345  -4518.345     e-Shop  Home and kitchen        Bath  
4       -2  -791  166.110  -1748.110   TeleShop             Books         DIY  
...    ...   ...      ...        ...        ...               ...         ...  
23048    1  1264  132.720   1396.720     e-Shop             Books    Academic  
23049    1   677   71.085    748.085     e-Shop          Clothing        Mens  
23050    4  1052  441.840   4649.840        MBR  Home and kitchen  Furnishing  
23051    3  1142  359.730   3785.730   TeleShop             Books    Children  
23052    1   447   46.935    493.935   TeleShop             Books    Children  

[23053 rows x 12 columns]

In [43]:
Customer_Final= pd.merge(left=tra_prod_data, right=customer_data,right_on="customer_Id", left_on="cust_id", how="left")
Customer_Final

transaction_id  cust_id   tran_date  prod_subcat_code  prod_cat_code  \
0         80712190438   270351  28-02-2014                 1              1   
1         29258453508   270384  27-02-2014                 5              3   
2         51750724947   273420  24-02-2014                 6              5   
3         93274880719   271509  24-02-2014                11              6   
4         51750724947   273420  23-02-2014                 6              5   
...               ...      ...         ...               ...            ...   
23048     94340757522   274550  25-01-2011                12              5   
23049     89780862956   270022  25-01-2011                 4              1   
23050     85115299378   271020  25-01-2011                 2              6   
23051     72870271171   270911  25-01-2011                11              5   
23052     77960931771   271961  25-01-2011                11              5   

       Qty  Rate      Tax  total_amt Store_type          prod_cat prod_subcat  \
0       -5  -772  405.300  -4265.300     e-Shop          Clothing       Women   
1       -5 -1497  785.925  -8270.925     e-Shop       Electronics   Computers   
2       -2  -791  166.110  -1748.110   TeleShop             Books         DIY   
3       -3 -1363  429.345  -4518.345     e-Shop  Home and kitchen        Bath   
4       -2  -791  166.110  -1748.110   TeleShop             Books         DIY   
...    ...   ...      ...        ...        ...               ...         ...   
23048    1  1264  132.720   1396.720     e-Shop             Books    Academic   
23049    1   677   71.085    748.085     e-Shop          Clothing        Mens   
23050    4  1052  441.840   4649.840        MBR  Home and kitchen  Furnishing   
23051    3  1142  359.730   3785.730   TeleShop             Books    Children   
23052    1   447   46.935    493.935   TeleShop             Books    Children   

       customer_Id         DOB Gender  city_code  
0           270351  26-09-1981      M        5.0  
1           270384  11-05-1973      F        8.0  
2           273420  27-07-1992      M        8.0  
3           271509  08-06-1981      M        3.0  
4           273420  27-07-1992      M        8.0  
...            ...         ...    ...        ...  
23048       274550  21-02-1972      M        7.0  
23049       270022  27-04-1984      M        9.0  
23050       271020  20-06-1976      M        8.0  
23051       270911  22-05-1970      M        2.0  
23052       271961  15-01-1982      M        1.0  

[23053 rows x 16 columns]

In [44]:
Customer_Final=Customer_Final.drop(['customer_Id'], axis=1)
Customer_Final

transaction_id  cust_id   tran_date  prod_subcat_code  prod_cat_code  \
0         80712190438   270351  28-02-2014                 1              1   
1         29258453508   270384  27-02-2014                 5              3   
2         51750724947   273420  24-02-2014                 6              5   
3         93274880719   271509  24-02-2014                11              6   
4         51750724947   273420  23-02-2014                 6              5   
...               ...      ...         ...               ...            ...   
23048     94340757522   274550  25-01-2011                12              5   
23049     89780862956   270022  25-01-2011                 4              1   
23050     85115299378   271020  25-01-2011                 2              6   
23051     72870271171   270911  25-01-2011                11              5   
23052     77960931771   271961  25-01-2011                11              5   

       Qty  Rate      Tax  total_amt Store_type          prod_cat prod_subcat  \
0       -5  -772  405.300  -4265.300     e-Shop          Clothing       Women   
1       -5 -1497  785.925  -8270.925     e-Shop       Electronics   Computers   
2       -2  -791  166.110  -1748.110   TeleShop             Books         DIY   
3       -3 -1363  429.345  -4518.345     e-Shop  Home and kitchen        Bath   
4       -2  -791  166.110  -1748.110   TeleShop             Books         DIY   
...    ...   ...      ...        ...        ...               ...         ...   
23048    1  1264  132.720   1396.720     e-Shop             Books    Academic   
23049    1   677   71.085    748.085     e-Shop          Clothing        Mens   
23050    4  1052  441.840   4649.840        MBR  Home and kitchen  Furnishing   
23051    3  1142  359.730   3785.730   TeleShop             Books    Children   
23052    1   447   46.935    493.935   TeleShop             Books    Children   

              DOB Gender  city_code  
0      26-09-1981      M        5.0  
1      11-05-1973      F        8.0  
2      27-07-1992      M        8.0  
3      08-06-1981      M        3.0  
4      27-07-1992      M        8.0  
...           ...    ...        ...  
23048  21-02-1972      M        7.0  
23049  27-04-1984      M        9.0  
23050  20-06-1976      M        8.0  
23051  22-05-1970      M        2.0  
23052  15-01-1982      M        1.0  

[23053 rows x 15 columns]

In [97]:
Customer_Final["tran_date"] = Customer_Final["tran_date"].astype('datetime64[ns]')
Customer_Final["DOB"] = Customer_Final["DOB"].astype('datetime64[ns]')

### 2. Prepare a summary report for the merged data set.

### a. Get the column names and their corresponding data types

In [98]:
Customer_Final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23053 entries, 0 to 23052
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    23053 non-null  int64         
 1   cust_id           23053 non-null  int64         
 2   tran_date         23053 non-null  datetime64[ns]
 3   prod_subcat_code  23053 non-null  int64         
 4   prod_cat_code     23053 non-null  int64         
 5   Qty               23053 non-null  int64         
 6   Rate              23053 non-null  int64         
 7   Tax               23053 non-null  float64       
 8   total_amt         23053 non-null  float64       
 9   Store_type        23053 non-null  object        
 10  prod_cat          23053 non-null  object        
 11  prod_subcat       23053 non-null  object        
 12  DOB               23053 non-null  datetime64[ns]
 13  Gender            23044 non-null  object        
 14  city_code         2304

### b. Top/Bottom 10 observations

In [46]:
Customer_Final.head(10)

transaction_id  cust_id   tran_date  prod_subcat_code  prod_cat_code  Qty  \
0     80712190438   270351  28-02-2014                 1              1   -5   
1     29258453508   270384  27-02-2014                 5              3   -5   
2     51750724947   273420  24-02-2014                 6              5   -2   
3     93274880719   271509  24-02-2014                11              6   -3   
4     51750724947   273420  23-02-2014                 6              5   -2   
5     97439039119   272357  23-02-2014                 8              3   -2   
6     45649838090   273667  22-02-2014                11              6   -1   
7     22643667930   271489  22-02-2014                12              6   -1   
8     79792372943   275108  22-02-2014                 3              1   -3   
9     50076728598   269014  21-02-2014                 8              3   -4   

   Rate      Tax  total_amt Store_type          prod_cat          prod_subcat  \
0  -772  405.300  -4265.300     e-Shop          Clothing                Women   
1 -1497  785.925  -8270.925     e-Shop       Electronics            Computers   
2  -791  166.110  -1748.110   TeleShop             Books                  DIY   
3 -1363  429.345  -4518.345     e-Shop  Home and kitchen                 Bath   
4  -791  166.110  -1748.110   TeleShop             Books                  DIY   
5  -824  173.040  -1821.040   TeleShop       Electronics  Personal Appliances   
6 -1450  152.250  -1602.250     e-Shop  Home and kitchen                 Bath   
7 -1225  128.625  -1353.625   TeleShop  Home and kitchen                Tools   
8  -908  286.020  -3010.020        MBR          Clothing                 Kids   
9  -581  244.020  -2568.020     e-Shop       Electronics  Personal Appliances   

          DOB Gender  city_code  
0  26-09-1981      M        5.0  
1  11-05-1973      F        8.0  
2  27-07-1992      M        8.0  
3  08-06-1981      M        3.0  
4  27-07-1992      M        8.0  
5  09-10-1982      F        6.0  
6  29-05-1981      M        9.0  
7  21-04-1971      M        9.0  
8  04-11-1971      F        8.0  
9  27-11-1979      F        3.0

In [47]:
Customer_Final.tail()

transaction_id  cust_id   tran_date  prod_subcat_code  prod_cat_code  \
23048     94340757522   274550  25-01-2011                12              5   
23049     89780862956   270022  25-01-2011                 4              1   
23050     85115299378   271020  25-01-2011                 2              6   
23051     72870271171   270911  25-01-2011                11              5   
23052     77960931771   271961  25-01-2011                11              5   

       Qty  Rate      Tax  total_amt Store_type          prod_cat prod_subcat  \
23048    1  1264  132.720   1396.720     e-Shop             Books    Academic   
23049    1   677   71.085    748.085     e-Shop          Clothing        Mens   
23050    4  1052  441.840   4649.840        MBR  Home and kitchen  Furnishing   
23051    3  1142  359.730   3785.730   TeleShop             Books    Children   
23052    1   447   46.935    493.935   TeleShop             Books    Children   

              DOB Gender  city_code  
23048  21-02-1972      M        7.0  
23049  27-04-1984      M        9.0  
23050  20-06-1976      M        8.0  
23051  22-05-1970      M        2.0  
23052  15-01-1982      M        1.0

### c. “Five-number summary” for continuous variables (min, Q1, median, Q3 and max)

In [48]:
Customer_Final.describe()

transaction_id        cust_id  prod_subcat_code  prod_cat_code  \
count    2.305300e+04   23053.000000      23053.000000   23053.000000   
mean     5.007348e+10  271021.746497          6.149091       3.763632   
std      2.898194e+10    2431.692059          3.726372       1.677016   
min      3.268991e+06  266783.000000          1.000000       1.000000   
25%      2.493864e+10  268935.000000          3.000000       2.000000   
50%      5.009313e+10  270980.000000          5.000000       4.000000   
75%      7.533000e+10  273114.000000         10.000000       5.000000   
max      9.998755e+10  275265.000000         12.000000       6.000000   

                Qty          Rate           Tax     total_amt     city_code  
count  23053.000000  23053.000000  23053.000000  23053.000000  23045.000000  
mean       2.432395    636.369713    248.667192   2107.308002      5.482534  
std        2.268406    622.363498    187.177773   2507.561264      2.863499  
min       -5.000000  -1499.000000      7.350000  -8270.925000      1.000000  
25%        1.000000    312.000000     98.280000    762.450000      3.000000  
50%        3.000000    710.000000    199.080000   1754.740000      5.000000  
75%        4.000000   1109.000000    365.715000   3569.150000      8.000000  
max        5.000000   1500.000000    787.500000   8287.500000     10.000000

In [49]:
quartiles = percentile(Customer_Final['total_amt'],[25,50,75])
data_min, data_max = Customer_Final['total_amt'].min(), Customer_Final['total_amt'].max()
# print 5-number summary
print('Min: %.3f' % data_min)
print('Q1: %.3f' % quartiles[0])
print('Median: %.3f' % quartiles[1])
print('Q3: %.3f' % quartiles[2])
print('Max: %.3f' % data_max)

Min: -8270.925
Q1: 762.450
Median: 1754.740
Q3: 3569.150
Max: 8287.500


In [50]:
quartiles = percentile(Customer_Final['Tax'] , [25, 50, 75])
data_min, data_max = Customer_Final['Tax'].min(), Customer_Final['Tax'].max()
# print 5-number summary
print('Min: %.3f' % data_min)
print('Q1: %.3f' % quartiles[0])
print('Median: %.3f' % quartiles[1])
print('Q3: %.3f' % quartiles[2])
print('Max: %.3f' % data_max)

Min: 7.350
Q1: 98.280
Median: 199.080
Q3: 365.715
Max: 787.500


In [51]:
quartiles = percentile(Customer_Final['Rate'] , [25, 50, 75])
data_min, data_max = Customer_Final['Rate'].min(), Customer_Final['Rate'].max()
# print 5-number summary
print('Min: %.3f' % data_min)
print('Q1: %.3f' % quartiles[0])
print('Median: %.3f' % quartiles[1])
print('Q3: %.3f' % quartiles[2])
print('Max: %.3f' % data_max)

Min: -1499.000
Q1: 312.000
Median: 710.000
Q3: 1109.000
Max: 1500.000


In [52]:
quartiles = percentile(Customer_Final['Qty'] , [25, 50, 75])
data_min, data_max = Customer_Final['Qty'].min(), Customer_Final['Tax'].max()
# print 5-number summary
print('Min: %.3f' % data_min)
print('Q1: %.3f' % quartiles[0])
print('Median: %.3f' % quartiles[1])
print('Q3: %.3f' % quartiles[2])
print('Max: %.3f' % data_max)

Min: -5.000
Q1: 1.000
Median: 3.000
Q3: 4.000
Max: 787.500


### d. Frequency tables for all the categorical variables

In [70]:
customer_category= Customer_Final.loc[:,Customer_Final.dtypes=='object']

In [71]:
customer_category.describe()

Store_type prod_cat prod_subcat         DOB Gender
count       23053    23053       23053       23053  23044
unique          4        6          18        3987      2
top        e-Shop    Books       Women  27-12-1988      M
freq         9311     6069        3048          32  11811

## 3. Generate histograms for all continuous variables and frequency bars for categorical variables.

In [72]:
Customer_Final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23053 entries, 0 to 23052
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   transaction_id    23053 non-null  int64         
 1   cust_id           23053 non-null  int64         
 2   tran_date         23053 non-null  datetime64[ns]
 3   prod_subcat_code  23053 non-null  int64         
 4   prod_cat_code     23053 non-null  int64         
 5   Qty               23053 non-null  int64         
 6   Rate              23053 non-null  int64         
 7   Tax               23053 non-null  float64       
 8   total_amt         23053 non-null  float64       
 9   Store_type        23053 non-null  object        
 10  prod_cat          23053 non-null  object        
 11  prod_subcat       23053 non-null  object        
 12  DOB               23053 non-null  object        
 13  Gender            23044 non-null  object        
 14  city_code         2304

In [83]:
Customer_Final['prod_subcat_code'].iplot(kind='hist',xTitle='Product Sub Category',yTitle='Frequency',title='Product Sub Category')

In [84]:
Customer_Final['prod_cat_code'].iplot(kind='hist',xTitle='Product Category',yTitle='Frequency',title='Product Category')

In [85]:
Customer_Final['Qty'].iplot(kind='hist',xTitle='Quantity',yTitle='Frequency',title='Quantity')

In [86]:
Customer_Final['Rate'].iplot(kind='hist',xTitle='Rate',yTitle='Frequency',title='Rate')

In [87]:
Customer_Final['Tax'].iplot(kind='hist',xTitle='Tax',yTitle='Frequency',title='Tax')

In [88]:
Customer_Final['total_amt'].iplot(kind='hist',xTitle='Total Amount',yTitle='Frequency',title='Toatal Amount')

## Bar chart of categorical variables

In [89]:
customer_category.columns

Index(['Store_type', 'prod_cat', 'prod_subcat', 'DOB', 'Gender'], dtype='object')

In [92]:
Customer_Final['Store_type'].value_counts().iplot(kind='bar',bins=100,xTitle='Store Type',linecolor='black',yTitle='Frequency',title='Store Type')

In [93]:
Customer_Final['prod_cat'].value_counts().iplot(kind='bar',bins=100,xTitle='Product Category',linecolor='black',yTitle='Frequency',title='Product Category')

In [94]:
Customer_Final['prod_subcat'].value_counts().iplot(kind='bar',bins=100,xTitle='Product Sub Category',linecolor='black',yTitle='Frequency',title='Product Sub Category')

In [95]:
Customer_Final['Gender'].value_counts().iplot(kind='bar',bins=100,xTitle='Gender Category',linecolor='black',yTitle='Frequency',title='Gender Category')

## 4. Calculate the following information using the merged dataset :

### a. Time period of the available transaction data

In [99]:
Customer_Final.columns

Index(['transaction_id', 'cust_id', 'tran_date', 'prod_subcat_code',
       'prod_cat_code', 'Qty', 'Rate', 'Tax', 'total_amt', 'Store_type',
       'prod_cat', 'prod_subcat', 'DOB', 'Gender', 'city_code'],
      dtype='object')

In [104]:
min_date= Customer_Final['tran_date'].min()
min_date=pd.Timestamp.strftime(min_date,format="%d-%m-%Y")
max_date= Customer_Final['tran_date'].max()
max_date=pd.Timestamp.strftime(max_date,format="%d-%m-%Y")
print("Time period of the available transaction data is From "+ min_date + " to " + max_date)

Time period of the available transaction data is From 02-01-2011 to 02-12-2014


### b. Count of transactions where the total amount of transaction was negative

In [106]:
Customer_Final[Customer_Final['total_amt'] < 0].value_counts().sum()

2176

## 5. Analyze which product categories are more popular among females vs male customers.

In [132]:
Customer_Final.columns

Index(['transaction_id', 'cust_id', 'tran_date', 'prod_subcat_code',
       'prod_cat_code', 'Qty', 'Rate', 'Tax', 'total_amt', 'Store_type',
       'prod_cat', 'prod_subcat', 'DOB', 'Gender', 'city_code'],
      dtype='object')

In [137]:
Prod_count1= Customer_Final.groupby(['prod_cat','Gender'])[['Qty']].sum().reset_index()
Prod_count1

prod_cat Gender   Qty
0               Bags      F  2362
1               Bags      M  2346
2              Books      F  7070
3              Books      M  7587
4           Clothing      F  3425
5           Clothing      M  3748
6        Electronics      F  5832
7        Electronics      M  6476
8           Footwear      F  3716
9           Footwear      M  3555
10  Home and kitchen      F  4895
11  Home and kitchen      M  5041

In [138]:
fig= px.bar(Prod_count1, x='prod_cat', y='Qty', color='Gender', barmode='group', height=400)
fig.show()

### Inference:

which product categories are more popular among females vs male customers are Books.

## 6. Which City code has the maximum customers and what was the percentage of customers from that city?

In [140]:
max_cust= customer_data.groupby(['city_code']).size().reset_index().rename(columns={0:'Customer Count'})
max_cust.sort_values(by='Customer Count', ascending=False)

city_code  Customer Count
2        3.0             595
4        5.0             586
3        4.0             582
6        7.0             575
1        2.0             564
7        8.0             562
9       10.0             558
0        1.0             551
8        9.0             546
5        6.0             526

In [141]:
fig = px.pie(max_cust, values='Customer Count', names='city_code', title='percentage of customers from the city')
fig.show()

### Inference:

City code which has the maximum customers and the percentage of customers from that city is city_code with 595 customer and percentage is 10.5%.

## 7. Which store type sells the maximum products by value and by quantity?

In [155]:
Store_sale= Customer_Final.groupby("Store_type", as_index=False)["Qty","Rate"].sum().sort_values(by="Qty",ascending=False)
Store_sale

C:\Users\rahul\AppData\Local\Temp\ipykernel_20000\1881129964.py:1: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



Store_type    Qty     Rate
3          e-Shop  22763  5941001
1             MBR  11194  2953253
0  Flagship store  11133  2941557
2        TeleShop  10984  2834420

### Inference:

store type that sells the maximum products by value and by quantity is shown in the above chart.

## 8. What was the total amount earned from the "Electronics" and "Clothing" categories from Flagship Stores?

In [164]:
Store_sale2= Customer_Final[Customer_Final['Store_type']=='Flagship store'].groupby("prod_cat")[['total_amt']].sum()
Store_sale2.loc[["Clothing","Electronics"]]

prod_cat
Clothing       1194423.23
Electronics    2215136.04
Name: total_amt, dtype: float64

In [169]:
Store_sale2.loc[["Clothing","Electronics"]].sum()

3409559.27

### Inference:

the total amount earned from the "Electronics" and "Clothing" categories from Flagship Stores is 3409559.27.

## 9. What was the total amount earned from "Male" customers under the "Electronics" category?

In [168]:
Store_sale3= Customer_Final[Customer_Final['prod_cat']=='Electronics'].groupby("Gender")[['total_amt']].sum()
Store_sale3.loc[['M']]

Gender
M    5703109.425
Name: total_amt, dtype: float64

### Inference:

the total amount earned from "Male" customers under the "Electronics" category is 5703109.425.

## 10. How many customers have more than 10 unique transactions, after removing all transactions which have any negative amounts?

In [171]:
data2= Customer_Final[Customer_Final['total_amt'] > 0]
data2

transaction_id  cust_id  tran_date  prod_subcat_code  prod_cat_code  \
10        29258453508   270384 2014-02-20                 5              3   
11        25455265351   267750 2014-02-20                12              6   
12         1571002198   275023 2014-02-20                 6              5   
14        36554696014   269345 2014-02-20                 3              5   
15        56814940239   268799 2014-02-20                 7              5   
...               ...      ...        ...               ...            ...   
23048     94340757522   274550 2011-01-25                12              5   
23049     89780862956   270022 2011-01-25                 4              1   
23050     85115299378   271020 2011-01-25                 2              6   
23051     72870271171   270911 2011-01-25                11              5   
23052     77960931771   271961 2011-01-25                11              5   

       Qty  Rate      Tax  total_amt Store_type          prod_cat prod_subcat  \
10       5  1497  785.925   8270.925     e-Shop       Electronics   Computers   
11       3  1360  428.400   4508.400     e-Shop  Home and kitchen       Tools   
12       4   587  246.540   2594.540     e-Shop             Books         DIY   
14       3  1253  394.695   4153.695     e-Shop             Books      Comics   
15       5   368  193.200   2033.200     e-Shop             Books     Fiction   
...    ...   ...      ...        ...        ...               ...         ...   
23048    1  1264  132.720   1396.720     e-Shop             Books    Academic   
23049    1   677   71.085    748.085     e-Shop          Clothing        Mens   
23050    4  1052  441.840   4649.840        MBR  Home and kitchen  Furnishing   
23051    3  1142  359.730   3785.730   TeleShop             Books    Children   
23052    1   447   46.935    493.935   TeleShop             Books    Children   

             DOB Gender  city_code  
10    1973-11-05      F        8.0  
11    1986-10-13      M        1.0  
12    1971-09-03      M        6.0  
14    1970-06-26      F       10.0  
15    1979-06-27      M        9.0  
...          ...    ...        ...  
23048 1972-02-21      M        7.0  
23049 1984-04-27      M        9.0  
23050 1976-06-20      M        8.0  
23051 1970-05-22      M        2.0  
23052 1982-01-15      M        1.0  

[20876 rows x 15 columns]

In [187]:
data3=data2.groupby(['cust_id'], as_index=False)[['transaction_id']].count()
data3[data3['transaction_id']>10]

cust_id  transaction_id
4      266794              11
2426   270535              11
2606   270803              11
3867   272741              11
4032   273014              11
4815   274227              11

In [190]:
data3[data3['transaction_id']>10].count()

cust_id           6
transaction_id    6
dtype: int64

### Inference:

customers have more than 10 unique transactions, after removing all transactions which have any negative amounts is 6.

## 11. For all customers aged between 25 - 35, find out:

### a. What was the total amount spent for “Electronics” and “Books” product categories?

In [193]:
def age(birthdate):
    today = date.today()
    age = today.year - birthdate.year - ((today.month, today.day) < (birthdate.month, birthdate.day))
    return age

In [195]:
Customer_Final['Age']=Customer_Final['DOB'].apply(age)
Customer_Final.head(2)

transaction_id  cust_id  tran_date  prod_subcat_code  prod_cat_code  \
0         80712190438   270351 2014-02-28                 1              1   
1         29258453508   270384 2014-02-27                 5              3   
2         51750724947   273420 2014-02-24                 6              5   
3         93274880719   271509 2014-02-24                11              6   
4         51750724947   273420 2014-02-23                 6              5   
...               ...      ...        ...               ...            ...   
23048     94340757522   274550 2011-01-25                12              5   
23049     89780862956   270022 2011-01-25                 4              1   
23050     85115299378   271020 2011-01-25                 2              6   
23051     72870271171   270911 2011-01-25                11              5   
23052     77960931771   271961 2011-01-25                11              5   

       Qty  Rate      Tax  total_amt Store_type          prod_cat prod_subcat  \
0       -5  -772  405.300  -4265.300     e-Shop          Clothing       Women   
1       -5 -1497  785.925  -8270.925     e-Shop       Electronics   Computers   
2       -2  -791  166.110  -1748.110   TeleShop             Books         DIY   
3       -3 -1363  429.345  -4518.345     e-Shop  Home and kitchen        Bath   
4       -2  -791  166.110  -1748.110   TeleShop             Books         DIY   
...    ...   ...      ...        ...        ...               ...         ...   
23048    1  1264  132.720   1396.720     e-Shop             Books    Academic   
23049    1   677   71.085    748.085     e-Shop          Clothing        Mens   
23050    4  1052  441.840   4649.840        MBR  Home and kitchen  Furnishing   
23051    3  1142  359.730   3785.730   TeleShop             Books    Children   
23052    1   447   46.935    493.935   TeleShop             Books    Children   

             DOB Gender  city_code  Age  
0     1981-09-26      M        5.0   40  
1     1973-11-05      F        8.0   48  
2     1992-07-27      M        8.0   29  
3     1981-08-06      M        3.0   40  
4     1992-07-27      M        8.0   29  
...          ...    ...        ...  ...  
23048 1972-02-21      M        7.0   50  
23049 1984-04-27      M        9.0   38  
23050 1976-06-20      M        8.0   45  
23051 1970-05-22      M        2.0   52  
23052 1982-01-15      M        1.0   40  

[23053 rows x 16 columns]

In [198]:
Customer_Final['Pass_Status']= np.logical_and(Customer_Final['Age']>25, Customer_Final['Age']<35)
Customer_Final.head(2)

transaction_id  cust_id  tran_date  prod_subcat_code  prod_cat_code  \
0         80712190438   270351 2014-02-28                 1              1   
1         29258453508   270384 2014-02-27                 5              3   
2         51750724947   273420 2014-02-24                 6              5   
3         93274880719   271509 2014-02-24                11              6   
4         51750724947   273420 2014-02-23                 6              5   
...               ...      ...        ...               ...            ...   
23048     94340757522   274550 2011-01-25                12              5   
23049     89780862956   270022 2011-01-25                 4              1   
23050     85115299378   271020 2011-01-25                 2              6   
23051     72870271171   270911 2011-01-25                11              5   
23052     77960931771   271961 2011-01-25                11              5   

       Qty  Rate      Tax  total_amt Store_type          prod_cat prod_subcat  \
0       -5  -772  405.300  -4265.300     e-Shop          Clothing       Women   
1       -5 -1497  785.925  -8270.925     e-Shop       Electronics   Computers   
2       -2  -791  166.110  -1748.110   TeleShop             Books         DIY   
3       -3 -1363  429.345  -4518.345     e-Shop  Home and kitchen        Bath   
4       -2  -791  166.110  -1748.110   TeleShop             Books         DIY   
...    ...   ...      ...        ...        ...               ...         ...   
23048    1  1264  132.720   1396.720     e-Shop             Books    Academic   
23049    1   677   71.085    748.085     e-Shop          Clothing        Mens   
23050    4  1052  441.840   4649.840        MBR  Home and kitchen  Furnishing   
23051    3  1142  359.730   3785.730   TeleShop             Books    Children   
23052    1   447   46.935    493.935   TeleShop             Books    Children   

             DOB Gender  city_code  Age  Pass_Status  
0     1981-09-26      M        5.0   40        False  
1     1973-11-05      F        8.0   48        False  
2     1992-07-27      M        8.0   29         True  
3     1981-08-06      M        3.0   40        False  
4     1992-07-27      M        8.0   29         True  
...          ...    ...        ...  ...          ...  
23048 1972-02-21      M        7.0   50        False  
23049 1984-04-27      M        9.0   38        False  
23050 1976-06-20      M        8.0   45        False  
23051 1970-05-22      M        2.0   52        False  
23052 1982-01-15      M        1.0   40        False  

[23053 rows x 17 columns]

In [202]:
cust_age= Customer_Final[Customer_Final['Pass_Status'] ==1]
cust_age.head(2)

transaction_id  cust_id  tran_date  prod_subcat_code  prod_cat_code  \
2         51750724947   273420 2014-02-24                 6              5   
4         51750724947   273420 2014-02-23                 6              5   
23        91116291703   268509 2014-02-20                 1              2   
28        88853694830   268444 2014-02-20                 4              4   
37        31384765864   267058 2014-02-19                 3              2   
...               ...      ...        ...               ...            ...   
23033     32887353269   268885 2011-01-25                 3              1   
23034     28972634039   275227 2011-01-25                 9              3   
23035     42161597232   267094 2011-01-25                 3              1   
23038     61812307223   274306 2011-01-25                 4              4   
23044     14787475597   273982 2011-01-25                 4              3   

       Qty  Rate      Tax  total_amt      Store_type     prod_cat prod_subcat  \
2       -2  -791  166.110  -1748.110        TeleShop        Books         DIY   
4       -2  -791  166.110  -1748.110        TeleShop        Books         DIY   
23       4  1243  522.060   5494.060             MBR     Footwear        Mens   
28      -3   -80   25.200   -265.200             MBR         Bags       Women   
37       1   793   83.265    876.265          e-Shop     Footwear       Women   
...    ...   ...      ...        ...             ...          ...         ...   
23033    1   276   28.980    304.980  Flagship store     Clothing        Kids   
23034    1   334   35.070    369.070             MBR  Electronics     Cameras   
23035    3   147   46.305    487.305             MBR     Clothing        Kids   
23038    4  1447  607.740   6395.740          e-Shop         Bags       Women   
23044    5   969  508.725   5353.725          e-Shop  Electronics     Mobiles   

             DOB Gender  city_code  Age  Pass_Status  
2     1992-07-27      M        8.0   29         True  
4     1992-07-27      M        8.0   29         True  
23    1989-08-17      M       10.0   32         True  
28    1992-02-01      F        6.0   30         True  
37    1992-06-02      F       10.0   30         True  
...          ...    ...        ...  ...          ...  
23033 1989-04-14      F        6.0   33         True  
23034 1989-09-23      M        4.0   32         True  
23035 1992-11-24      M        9.0   29         True  
23038 1989-12-01      M        2.0   32         True  
23044 1991-12-10      M        4.0   30         True  

[5682 rows x 17 columns]

In [216]:
Store_sale5= cust_age.groupby("prod_cat")['total_amt'].sum()
Store_sale5

prod_cat
Bags                 954134.350
Books               2974622.430
Clothing            1646641.165
Electronics         2594277.010
Footwear            1583911.420
Home and kitchen    2078264.110
Name: total_amt, dtype: float64

In [218]:
Store_sale5.loc[["Electronics","Books"]].sum()

5568899.4399999995

### Inference:

 the total amount spent for “Electronics” and “Books” product categories 5568899.44.

## b. What was the total amount spent by these customers between 1st Jan, 2014 to 1st Mar, 2014?

In [219]:
cust_age= cust_age[(cust_age['tran_date']> '2014-01-01') & (cust_age['tran_date'] < '2014-03-01')]
cust_age

transaction_id  cust_id  tran_date  prod_subcat_code  prod_cat_code  \
2        51750724947   273420 2014-02-24                 6              5   
4        51750724947   273420 2014-02-23                 6              5   
23       91116291703   268509 2014-02-20                 1              2   
28       88853694830   268444 2014-02-20                 4              4   
37       31384765864   267058 2014-02-19                 3              2   
...              ...      ...        ...               ...            ...   
1013     32899104170   270064 2014-02-01                 2              6   
1023     14302941720   273058 2014-02-01                 1              2   
1025     32125935023   272067 2014-02-01                10              3   
1026     50346649770   270616 2014-02-01                 4              2   
1029     91039584326   272165 2014-02-01                10              5   

      Qty  Rate      Tax  total_amt Store_type          prod_cat  \
2      -2  -791  166.110  -1748.110   TeleShop             Books   
4      -2  -791  166.110  -1748.110   TeleShop             Books   
23      4  1243  522.060   5494.060        MBR          Footwear   
28     -3   -80   25.200   -265.200        MBR              Bags   
37      1   793   83.265    876.265     e-Shop          Footwear   
...   ...   ...      ...        ...        ...               ...   
1013    3   802  252.630   2658.630   TeleShop  Home and kitchen   
1023    1   674   70.770    744.770     e-Shop          Footwear   
1025    5  1211  635.775   6690.775     e-Shop       Electronics   
1026    2   977  205.170   2159.170        MBR          Footwear   
1029    1  1268  133.140   1401.140     e-Shop             Books   

          prod_subcat        DOB Gender  city_code  Age  Pass_Status  
2                 DIY 1992-07-27      M        8.0   29         True  
4                 DIY 1992-07-27      M        8.0   29         True  
23               Mens 1989-08-17      M       10.0   32         True  
28              Women 1992-02-01      F        6.0   30         True  
37              Women 1992-06-02      F       10.0   30         True  
...               ...        ...    ...        ...  ...          ...  
1013       Furnishing 1990-02-22      F        3.0   32         True  
1023             Mens 1989-11-08      F       10.0   32         True  
1025  Audio and video 1989-05-14      M        5.0   33         True  
1026             Kids 1988-01-08      M        9.0   34         True  
1029      Non-Fiction 1990-01-12      M        4.0   32         True  

[150 rows x 17 columns]

In [223]:
cust_age2=cust_age.groupby('cust_id', as_index=False)[['total_amt']].sum()
cust_age2

cust_id  total_amt
0     266813   1037.595
1     266841   2648.685
2     266853   2545.920
3     266987   6302.920
4     267058   3822.195
..       ...        ...
128   274727    434.265
129   274799   8154.900
130   274923   4486.300
131   274984   2789.020
132   275215      0.000

[133 rows x 2 columns]

In [225]:
cust_age2.total_amt.sum()

370927.505

### Inference:

the total amount spent by these customers between 1st Jan, 2014 to 1st Mar, 2014 370927.505.